### PR TITLE
vectorstores: Add remote chroma "vectorstore" adapter

### PR DESCRIPTION
**Pull request by noodnik2**
## implement [[Feature Request] able to use chromadb as the vector stores](https://github.com/tmc/langchaingo/issues/278) 

Proposed implementation of (remote) access to [Chroma](https://www.trychroma.com/) via [chroma-go](https://github.com/amikos-tech/chroma-go) library.  

### Eyes 👀 Wanted!

You can help this initiative by reviewing the proposed changes and leaving in-line PR comments!

### Scope / Success Criteria

The ambitions for the initial proposal are modest: create a "kernel" of support for Chrome that can be extended as new use cases or edge cases are discovered.

The initial test cases are based upon those already in place for the "Pinecone" adapter.  Additional tests are forthcoming for increased test coverage and new functionality.

### Motivations

My personal motivation for this stems from dual loyalties: 
- As an AI enthusiast / developer I've enjoyed the convenience and lack of "lock in" provided by Chroma from its "in-process" and "in-docker" access to localized vector storage for development workflows / experimentation; yet,
- As a Gopher, I've felt a little left out of the "LangChain" party, until now 😄  

### Due Diligence
Investigation into the various ways to interface LangChain directly with Chroma has led to the following conclusions:
- No "in-process" implementation yet exists for `golang`; therefore, initial access is being offered (only) to an (external) Chroma service.
- Discussions in [`langchaingo`'s Discord channel](https://discord.com/channels/1038097195422978059/1097608449266831390) and notes found in [chroma-core#1072](https://github.com/chroma-core/chroma/issues/1072) have suggested the choice to use the [chroma-go](https://github.com/amikos-tech/chroma-go) library's APIs to interact with the Chroma service.

### PR Checklist

- [X] Read the [Contributing documentation](https://github.com/tmc/langchaingo/blob/main/CONTRIBUTING.md).
- [X] Read the [Code of conduct documentation](https://github.com/tmc/langchaingo/blob/main/CODE_OF_CONDUCT.md).
- [X] Name your Pull Request title clearly, concisely, and prefixed with the name of the primarily affected package you changed according to [Good commit messages](https://go.dev/doc/contribute#commit_messages) (such as `memory: add interfaces for X, Y` or `util: add whizzbang helpers`).
- [X] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [X] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `Fixes #123`).
- [X] Describes the source of new concepts.
- [X] References existing implementations as appropriate.
- [X] Contains test coverage for new functions.
- [x] Passes all [`golangci-lint`](https://golangci-lint.run/) checks.

### CODE REVIEW

**Comment by noodnik2** [comment]
Please see [related Discord thread](https://discord.com/channels/1038097195422978059/1149033854452109412).

File: vectorstores/chroma/chroma.go [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+	ErrUnsupportedOptions       = errors.New("unsupported options")
+)
+
+// Store is a wrapper around the chromaGo API and client.
+type Store struct {
+	client           *chromago.Client
+	collection       *chromago.Collection
+	distanceFunction chromago.DistanceFunction
+	chromaURL        string
+	openaiAPIKey     string
```
Comment (bjwswang) [inline vectorstores/chroma/chroma.go:32]: what is this apikey for in this Store?

Comment (noodnik2) [reply vectorstores/chroma/chroma.go:32]: The "Open API Key" is the key used to interact with the OpenAI LLM.

**Comment by noodnik2** [comment]
Hi @tmc, have you seen this and the related collaboration @bjwswang and I have been having here and in Discord?

Any thoughts or guidance?

File: internal/util/util.go [hunk 2]
```diff
@@ -39,3 +39,13 @@ func ListKeys[T any](m map[string]T) []string {
 	}
 	return keys
 }
+
+// NewMap will create a new map based upon the provided map.
+// If the provided map is `nil`, a new, empty map will be returned.
+func NewMap[K, V comparable](m map[K]V) map[K]V {
```
Comment (tmc) [inline internal/util/util.go:45]: why this over https://pkg.go.dev/golang.org/x/exp/maps#Copy ?

File: vectorstores/chroma/chroma.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+		embeddingFunction, s.distanceFunction)
+	if errCc != nil {
+		return s, fmt.Errorf("%w: %w", ErrNewClient, errCc)
+	}
+
+	s.collection = col
+	return s, nil
+}
+
+func (s Store) AddDocuments(_ context.Context, docs []schema.Document, options ...vectorstores.Option) error {
```
Comment (tmc) [inline vectorstores/chroma/chroma.go:76]: comment exported methods

File: internal/util/util.go (hunk 2 above)
Comment (noodnik2) [reply internal/util/util.go:45]: Fair enough; I just learned about that; will do - thanks!

**Comment by noodnik2** [comment]
Thanks for the feedback, @tmc .  Now that I've got "a real job," my velocity has slowed down, but I do intend to continue to move this forward at (hopefully) a "reasonable pace."  In fact, there's possibly an (additional) application I've got in mind in my new role!

File: vectorstores/chroma/chroma.go (hunk 3 above)
Comment (noodnik2) [reply vectorstores/chroma/chroma.go:77]: Comment added.

**Comment by eliben** [comment]
Note that this PR adds a 15 MiB binary file (`chroma_vectorstore_example`) to the root of the directory. ACAICT this binary file isn't referenced anywhere in the code. Is this by mistake, or am I missing something?
